shotover-proxy/src/codec/cassandra.rs

```diff
@@ -67,96 +153,267 @@ impl CassandraDecoder {
+
+                    if header_crc24 != computed_crc {
+                        return Err(anyhow!(format!(
+                            "Header CRC mismatch - expected {header_crc24}, found {computed_crc}."
+                        )));
+                    }
+
+                    let payload_length = (header & 0x1ffff) as usize;
+                    let payload_end = UNCOMPRESSED_FRAME_HEADER_LENGTH + payload_length;
+
+                    let frame_end = payload_end + FRAME_TRAILER_LENGTH;
+
+                    let payload_crc32 =
+                        u32::from_le_bytes(frame_bytes[payload_end..frame_end].try_into().unwrap());
+
+                    let computed_crc = cassandra_protocol::crc::crc32(
+                        &frame_bytes[UNCOMPRESSED_FRAME_HEADER_LENGTH..payload_end],
                     );
+                    if payload_crc32 != computed_crc {
+                        return Err(anyhow!(format!(
+                            "Payload CRC mismatch - read {payload_crc32}, computed {computed_crc}."
+                        )));
+                    }
 
-                    let version = Version::try_from(bytes[0])
-                        .expect("Gauranteed because check_envelope_size only returns Ok if the Version will parse");
-                    if let Version::V3 | Version::V4 = version {
-                        // Accept these protocols
-                    } else {
-                        // Reject protocols that cassandra-protocol supports but shotover does not yet support
-                        return Err(reject_protocol_version(version.into()));
+                    let self_contained = (header & (1 << 17)) != 0;
+                    if !self_contained {
+                        unimplemented!("Cannot support non-self contained frames yet");
                     }
 
-                    let compressed = self.check_compression(&bytes).unwrap();
+                    frame_bytes.advance(UNCOMPRESSED_FRAME_HEADER_LENGTH);
+                    let mut payload = frame_bytes.split_to(payload_length);
 
-                    let mut message = Message::from_bytes(
-                        bytes.freeze(),
-                        crate::message::ProtocolType::Cassandra {
-                            compression: if compressed {
-                                *self.compression.read().unwrap()
-                            } else {
-                                Compression::None
+                    let mut envelopes: Vec<Message> = vec![];
+
+                    loop {
+                        if payload.len() < ENVELOPE_HEADER_LEN {
+                            break;
+                        }
+
+                        let body_len =
+                            i32::from_be_bytes(payload[5..9].try_into().unwrap()) as usize;
+
+                        let envelope_len = ENVELOPE_HEADER_LEN + body_len;
+
+                        if envelope_len > payload.len() {
+                            break;
```

Comment:
This should be treated as an error as its a violation of the cassandra protocol for the body of a frame to be longer than the bytes of the message.
Also it should probably be:
```rust
if envelope_len > payload.len() - CRC32_LEN {
```